File: frontend/src/lib/report_builder.ts
const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/reports-builder`;

async function req<T>(path: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...opts,
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? r.statusText);
  }
  if (r.status === 204) return undefined as T;
  return r.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type ReportVisibility = "private" | "team" | "global";
export type AggregationType = "none" | "sum" | "avg" | "min" | "max" | "count" | "count_distinct";
export type FilterOperator = "eq" | "neq" | "gt" | "lt" | "gte" | "lte" | "between" | "in" | "like" | "is_null" | "is_not_null";
export type LogicalOperator = "and" | "or";
export type SortDirection = "asc" | "desc";
export type ChartType = "table" | "bar" | "line" | "pie" | "kpi" | "area" | "scatter";
export type ScheduleFrequency = "daily" | "weekly" | "monthly" | "custom";
export type ExportFormat = "csv" | "xlsx" | "json";
export type RBAIAgentType = "builder_assistant" | "insight_generator" | "performance_optimizer";
export type RBAIRecStatus = "pending" | "acknowledged" | "actioned" | "dismissed";

export interface DataSourceField {
  path: string;
  label: string;
  type: "string" | "number" | "date" | "boolean" | "datetime";
}

export interface DataSource {
  label: string;
  module: string;
  field_count: number;
  fields: DataSourceField[];
}

export interface ReportField {
  report_field_id: string;
  report_id: string;
  field_path: string;
  alias: string | null;
  data_type: string;
  aggregation: AggregationType;
  is_group_by: boolean;
  sort_direction: SortDirection | null;
  sort_order: number | null;
  position: number;
  visible: boolean;
}

export interface ReportFilter {
  filter_id: string;
  report_id: string;
  field_path: string;
  operator: FilterOperator;
  value: string | null;
  value_to: string | null;
  logical_op: LogicalOperator;
  position: number;
}

export interface CalcField {
  calc_id: string;
  report_id: string;
  expression: string;
  alias: string;
  data_type: string;
  description: string | null;
}

export interface Schedule {
  schedule_id: string;
  report_id: string;
  frequency: ScheduleFrequency;
  run_time: string | null;
  recipients: string | null;
  export_format: ExportFormat;
  active_flag: boolean;
  last_run_at: string | null;
  next_run_at: string | null;
  created_at: string;
}

export interface Report {
  report_id: string;
  report_code: string;
  report_name: string;
  description: string | null;
  owner_user_id: string | null;
  owner_name: string | null;
  visibility: ReportVisibility;
  data_source: string;
  query_definition: Record<string, unknown> | null;
  active_flag: boolean;
  is_template: boolean;
  run_count: number;
  last_run_at: string | null;
  created_at: string;
  fields: ReportField[];
  filters: ReportFilter[];
  calculated_fields: CalcField[];
  schedules: Schedule[];
}

export interface RunResult {
  report_id: string;
  report_name: string;
  data_source: string;
  columns: { path: string; label: string; type: string }[];
  rows: Record<string, string | null>[];
  total_count: number;
  execution_ms: number;
  page: number;
  page_size: number;
}

export interface Widget {
  widget_id: string;
  dashboard_id: string;
  report_id: string | null;
  widget_title: string | null;
  chart_type: ChartType;
  position: number;
  width: number;
  height: number;
  x_axis_field: string | null;
  y_axis_field: string | null;
  color_field: string | null;
  refresh_interval: number | null;
  config: Record<string, unknown> | null;
  created_at: string;
}

export interface Dashboard {
  dashboard_id: string;
  dashboard_code: string;
  dashboard_name: string;
  description: string | null;
  owner_user_id: string | null;
  visibility: ReportVisibility;
  active_flag: boolean;
  created_at: string;
  widgets: Widget[];
}

export interface RBAIRec {
  rec_id: string;
  agent_type: RBAIAgentType;
  status: RBAIRecStatus;
  title: string;
  body: string;
  report_id: string | null;
  score: number | null;
  actioned_by: string | null;
  actioned_at: string | null;
  created_at: string;
}

// ── Label Maps ────────────────────────────────────────────────────────────────

export const AGG_LABEL: Record<AggregationType, string> = {
  none: "—", sum: "SUM", avg: "AVG", min: "MIN", max: "MAX",
  count: "COUNT", count_distinct: "COUNT DISTINCT",
};

export const OP_LABEL: Record<FilterOperator, string> = {
  eq: "=", neq: "≠", gt: ">", lt: "<", gte: "≥", lte: "≤",
  between: "Between", in: "In list", like: "Contains",
  is_null: "Is empty", is_not_null: "Is not empty",
};

export const CHART_LABEL: Record<ChartType, string> = {
  table: "Table", bar: "Bar Chart", line: "Line Chart",
  pie: "Pie Chart", kpi: "KPI Card", area: "Area Chart", scatter: "Scatter",
};

export const CHART_TYPES: ChartType[] = ["table","bar","line","pie","kpi","area"];
export const FILTER_OPS: FilterOperator[] = ["eq","neq","gt","lt","gte","lte","between","in","like","is_null","is_not_null"];
export const AGG_TYPES: AggregationType[] = ["none","sum","avg","min","max","count","count_distinct"];

// ── API ───────────────────────────────────────────────────────────────────────

export const reportBuilderApi = {
  getCatalog: () => req<Record<string, DataSource>>("/catalog"),
  getDataSource: (id: string) => req<DataSource>(`/catalog/${id}`),

  createReport: (d: object) => req<Report>("/reports", { method: "POST", body: JSON.stringify(d) }),
  listReports: (params?: { data_source?: string; owner?: string; template_only?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.data_source) q.set("data_source", params.data_source);
    if (params?.owner) q.set("owner", params.owner);
    if (params?.template_only) q.set("template_only", "true");
    return req<Report[]>(`/reports?${q}`);
  },
  getReport: (id: string) => req<Report>(`/reports/${id}`),
  updateReport: (id: string, d: object) => req<Report>(`/reports/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  deleteReport: (id: string) => req<void>(`/reports/${id}`, { method: "DELETE" }),
  cloneReport: (id: string, new_code: string, new_name: string) =>
    req<Report>(`/reports/${id}/clone?new_code=${encodeURIComponent(new_code)}&new_name=${encodeURIComponent(new_name)}`, { method: "POST" }),
  seedTemplates: () => req<{ created: number }>("/reports/seed-templates", { method: "POST" }),

  runReport: (id: string, d: object) => req<RunResult>(`/reports/${id}/run`, { method: "POST", body: JSON.stringify(d) }),
  getExportUrl: (id: string) => `${BASE}/reports/${id}/export`,

  preview: (data_source: string, limit = 5) =>
    req<{ columns: unknown[]; rows: unknown[]; total: number }>(`/preview?data_source=${data_source}&limit=${limit}`, { method: "POST" }),

  createSchedule: (report_id: string, d: object) =>
    req<Schedule>(`/reports/${report_id}/schedule`, { method: "POST", body: JSON.stringify(d) }),
  listSchedules: (report_id?: string) =>
    req<Schedule[]>(`/schedules${report_id ? `?report_id=${report_id}` : ""}`),
  deactivateSchedule: (id: string) => req<void>(`/schedules/${id}`, { method: "DELETE" }),

  createDashboard: (d: object) => req<Dashboard>("/dashboards", { method: "POST", body: JSON.stringify(d) }),
  listDashboards: () => req<Dashboard[]>("/dashboards"),
  getDashboard: (id: string) => req<Dashboard>(`/dashboards/${id}`),
  addWidget: (dashboard_id: string, d: object) =>
    req<Widget>(`/dashboards/${dashboard_id}/widgets`, { method: "POST", body: JSON.stringify(d) }),
  deleteWidget: (id: string) => req<void>(`/widgets/${id}`, { method: "DELETE" }),

  runBuilderAssistant: () => req<RBAIRec[]>("/ai/run-builder-assistant", { method: "POST" }),
  runInsightGenerator: () => req<RBAIRec[]>("/ai/run-insight-generator", { method: "POST" }),
  runPerformanceOptimizer: () => req<RBAIRec[]>("/ai/run-performance-optimizer", { method: "POST" }),
  listAIRecs: (status?: string) => req<RBAIRec[]>(`/ai/recommendations${status ? `?status=${status}` : ""}`),
  ackAIRec: (id: string, d: object) =>
    req<RBAIRec>(`/ai/recommendations/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
};

File: backend/app/schemas/report_builder.py
from __future__ import annotations
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel

from app.models.report_builder import (
    ReportVisibility, AggregationType, FilterOperator, LogicalOperator,
    SortDirection, ChartType, ScheduleFrequency, ExportFormat,
    RBAIAgentType, RBAIRecStatus,
)


# ── Field / Filter ────────────────────────────────────────────────────────────

class ReportFieldCreate(BaseModel):
    field_path: str
    alias: Optional[str] = None
    data_type: str = "string"
    aggregation: AggregationType = AggregationType.NONE
    is_group_by: bool = False
    sort_direction: Optional[SortDirection] = None
    sort_order: Optional[int] = None
    position: int = 0
    visible: bool = True


class ReportFieldOut(BaseModel):
    report_field_id: UUID
    report_id: UUID
    field_path: str
    alias: Optional[str]
    data_type: str
    aggregation: AggregationType
    is_group_by: bool
    sort_direction: Optional[SortDirection]
    sort_order: Optional[int]
    position: int
    visible: bool

    class Config:
        from_attributes = True


class ReportFilterCreate(BaseModel):
    field_path: str
    operator: FilterOperator
    value: Optional[str] = None
    value_to: Optional[str] = None
    logical_op: LogicalOperator = LogicalOperator.AND
    position: int = 0


class ReportFilterOut(BaseModel):
    filter_id: UUID
    report_id: UUID
    field_path: str
    operator: FilterOperator
    value: Optional[str]
    value_to: Optional[str]
    logical_op: LogicalOperator
    position: int

    class Config:
        from_attributes = True


class CalcFieldCreate(BaseModel):
    expression: str
    alias: str
    data_type: str = "number"
    description: Optional[str] = None


class CalcFieldOut(BaseModel):
    calc_id: UUID
    report_id: UUID
    expression: str
    alias: str
    data_type: str
    description: Optional[str]

    class Config:
        from_attributes = True


# ── Schedule ──────────────────────────────────────────────────────────────────

class ScheduleCreate(BaseModel):
    frequency: ScheduleFrequency
    run_time: Optional[str] = "08:00"
    recipients: Optional[str] = None
    export_format: ExportFormat = ExportFormat.CSV


class ScheduleOut(BaseModel):
    schedule_id: UUID
    report_id: UUID
    frequency: ScheduleFrequency
    run_time: Optional[str]
    recipients: Optional[str]
    export_format: ExportFormat
    active_flag: bool
    last_run_at: Optional[datetime]
    next_run_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


# ── Report ────────────────────────────────────────────────────────────────────

class ReportCreate(BaseModel):
    report_code: str
    report_name: str
    description: Optional[str] = None
    owner_user_id: Optional[str] = None
    owner_name: Optional[str] = None
    visibility: ReportVisibility = ReportVisibility.PRIVATE
    data_source: str
    fields: List[ReportFieldCreate] = []
    filters: List[ReportFilterCreate] = []
    calculated_fields: List[CalcFieldCreate] = []
    query_definition: Optional[Dict[str, Any]] = None


class ReportUpdate(BaseModel):
    report_name: Optional[str] = None
    description: Optional[str] = None
    visibility: Optional[ReportVisibility] = None
    data_source: Optional[str] = None
    query_definition: Optional[Dict[str, Any]] = None
    active_flag: Optional[bool] = None


class ReportOut(BaseModel):
    report_id: UUID
    report_code: str
    report_name: str
    description: Optional[str]
    owner_user_id: Optional[str]
    owner_name: Optional[str]
    visibility: ReportVisibility
    data_source: str
    query_definition: Optional[Dict[str, Any]]
    active_flag: bool
    is_template: bool
    run_count: int
    last_run_at: Optional[datetime]
    created_at: datetime
    fields: List[ReportFieldOut] = []
    filters: List[ReportFilterOut] = []
    calculated_fields: List[CalcFieldOut] = []
    schedules: List[ScheduleOut] = []

    class Config:
        from_attributes = True


class RunRequest(BaseModel):
    limit: int = 100
    offset: int = 0
    runtime_filters: Optional[Dict[str, str]] = None


class RunResult(BaseModel):
    report_id: UUID
    report_name: str
    data_source: str
    columns: List[Dict[str, str]]
    rows: List[Dict[str, Any]]
    total_count: int
    execution_ms: int
    page: int
    page_size: int


# ── Dashboard ─────────────────────────────────────────────────────────────────

class WidgetCreate(BaseModel):
    report_id: Optional[UUID] = None
    widget_title: Optional[str] = None
    chart_type: ChartType = ChartType.TABLE
    position: int = 0
    width: int = 6
    height: int = 4
    x_axis_field: Optional[str] = None
    y_axis_field: Optional[str] = None
    color_field: Optional[str] = None
    refresh_interval: Optional[int] = None
    config: Optional[Dict[str, Any]] = None


class WidgetOut(BaseModel):
    widget_id: UUID
    dashboard_id: UUID
    report_id: Optional[UUID]
    widget_title: Optional[str]
    chart_type: ChartType
    position: int
    width: int
    height: int
    x_axis_field: Optional[str]
    y_axis_field: Optional[str]
    color_field: Optional[str]
    refresh_interval: Optional[int]
    config: Optional[Dict[str, Any]]
    created_at: datetime

    class Config:
        from_attributes = True


class DashboardCreate(BaseModel):
    dashboard_code: str
    dashboard_name: str
    description: Optional[str] = None
    owner_user_id: Optional[str] = None
    visibility: ReportVisibility = ReportVisibility.TEAM


class DashboardOut(BaseModel):
    dashboard_id: UUID
    dashboard_code: str
    dashboard_name: str
    description: Optional[str]
    owner_user_id: Optional[str]
    visibility: ReportVisibility
    active_flag: bool
    created_at: datetime
    widgets: List[WidgetOut] = []

    class Config:
        from_attributes = True


# ── AI ────────────────────────────────────────────────────────────────────────

class RBAIRecOut(BaseModel):
    rec_id: UUID
    agent_type: RBAIAgentType
    status: RBAIRecStatus
    title: str
    body: str
    report_id: Optional[UUID]
    score: Optional[int]
    actioned_by: Optional[str]
    actioned_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class RBAIRecAck(BaseModel):
    status: RBAIRecStatus
    actioned_by: Optional[str] = None


class ScheduleRunLogRead(BaseModel):
    run_id: UUID
    schedule_id: UUID
    report_id: UUID
    started_at: datetime
    completed_at: Optional[datetime] = None
    status: str
    row_count: Optional[int] = None
    export_format: Optional[str] = None
    recipients_sent: Optional[str] = None
    error_message: Optional[str] = None

    class Config:
        from_attributes = True


ReportOut.model_rebuild()
DashboardOut.model_rebuild()
